Tests/Client/Custom Assertions/AssertNetworkRequest.swift

```diff
@@ -0,0 +1,150 @@
+    // Check path
+    if components.path != path {
+        errorMessage += "\n  - Incorrect path: expected \"\(path)\" got \"\(components.path)\""
+    }
+
+    // Check headers
+    let requestHeaders = request.allHTTPHeaderFields ?? [:]
+    headers?.forEach { (key, value) in
+        if let requestHeaderValue = requestHeaders[key] {
+            if value != requestHeaderValue {
+                errorMessage += "\n  - Incorrect header value for \"\(key)\": expected \"\(value)\" got \"\(requestHeaderValue)\""
+            }
+
+        } else {
+            errorMessage += "\n  - Missing header value for \"\(key)\""
+        }
+    }
+
+    // Check query parameters
+    let items = components.queryItems ?? []
+    queryParameters?.forEach { (key, value) in
+        if let requestValue = items[key] {
+            if value != requestValue {
+                errorMessage += "\n  - Incorrect query value for \"\(key)\": expected \"\(value)\" got \"\(requestValue)\""
+            }
+
+        } else {
+            errorMessage += "\n  - Missing query value for \"\(key)\""
+        }
+    }
+
+    // Check the request body
+    var requestBodyData: Data?
+    if let data = request.httpBody {
+        requestBodyData = data
+    }
+    if let stream = request.httpBodyStream {
+        requestBodyData = Data(reading: stream)
+    }
+
+    guard let data = requestBodyData else {
+        if body != nil {
+            errorMessage += "\n  - Missing request body"
+        }
+        return
+    }
+
+    guard let requestBodyJSON = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
+        errorMessage += "\n  - Request body is not a valid JSON"
+        return
+    }
+
+    body?.forEach { (key, value) in
+        if let requestValue = requestBodyJSON[key] {
+            if String(describing: value) != String(describing: requestValue) {
+                errorMessage += "\n  - Incorrect body value for \"\(key)\": expected \"\(value)\" got \"\(requestValue)\""
+            }
+        } else {
+            errorMessage += "\n  - Missing body value for \"\(key)\""
+        }
+    }
+}
+
+private extension Array where Element == URLQueryItem {
+  /// Returns the value of the URLQueryItem with the given name. Returns `nil`
+  /// if the query item doesn't exist.
+  subscript(_ name: String) -> String? {
+    self.first(where: { $0.name == name}).flatMap({ $0.value })
+  }
+}
+
+private extension Data {
+    /// Creates a new Data instance from the provided InputStream. It opens and closes the stream during the process.
+    init(reading input: InputStream) {
+        self.init()
+        input.open()
+
+        let bufferSize = 1024
+        let buffer = UnsafeMutablePointer<UInt8>.allocate(capacity: bufferSize)
+        while input.hasBytesAvailable {
+            let read = input.read(buffer, maxLength: bufferSize)
+            if (read == 0) {
+                break  // added
+            }
+            self.append(buffer, count: read)
```

Comment:
No need `self.`.
It's good to have this function, but currently, we don't use data from `InputStream`.

Reply:
- fixed
- it's used actually. the body of the tests in the PR comes from `request.httpBodyStream`. Is it unexpected @buh ?

Reply:
Yes, it's weird.